tensorboard/examples/plugins/example_raw_scalars/tensorboard_plugin_example_raw_scalars/plugin.py

```diff
@@ -84,10 +85,18 @@ def _serve_static_file(self, request):
         Requests from the frontend have a path in this form:
         /data/plugin/example_raw_scalars/static/foo
         This serves the appropriate asset: ./static/foo.
+
+        Checks the normpath to guard against path traversal attacks.
         """
         static_path_part = request.path[len(_PLUGIN_DIRECTORY_PATH_PART) :]
-        res_path = os.path.join(os.path.dirname(__file__), static_path_part)
+        posix_norm_path = posixpath.normpath(static_path_part)
+        if not posix_norm_path.startswith("static/"):
+            return http_util.Respond(
+                request, "Not found", "text/plain", code=404
+            )
 
+        norm_path = os.path.normpath(static_path_part)
```

Comment:
If I read this correctly, the vulnerability still exists on Windows with
a path like `static/%2e%2e%5c%2e%2e%5ctest.txt`. You’re checking for
safety of the POSIX-parsed path but then using the NT path resolver to
actually determine the path, so the thing that you’ve tested for safety
isn’t the thing that’s actually used:

```
>>> url = urllib.parse.unquote("static/%2e%2e%5c%2e%2e%5ctest.txt")
>>> print(posixpath.normpath(url))  # note: starts with "static/"
static/..\..\test.txt
>>> print(ntpath.join(r"C:\path\to\dir", ntpath.normpath(url)))
C:\path\to\dir\..\test.txt
```

Reply:
Yes, it was broken!  I realize now that normpath("a\\b\\c") doesn't convert to a posixpath "a/b/c".

I think what we need is to:
convert to posix path (no mixed slashes now) -->
posixpath.join with the current directory (closer to the path that `open` will call) -->
posixpath.normpath (any dot-dot-slash's should be at the beginning) -->
check that it starts with the right prefix.

That way, the result of the check can be the same thing we pass to `open`.

One thing I'm wondering is if `replace(ntpath.sep, posixpath.sep)` is the best way to ensure a path is a posixpath.